Cache-Hash-Data-Structure.adoc

```diff
@@ -0,0 +1,174 @@
+= Hash Data Structure
+
+Provide a Hash data structure to Infinispan users. The Hash is a flexible identifiable grouping of keys mapping to values.
+In this context, _flexible_ means it can have arbitrary mappings without needing a schema beforehand; _identifiable group_
+means the Hash structure receives a unique name and retrieves the complete key-value mapping.
+
+The Hash provides a general approach for users to represent their data structures using Infinispan. For example, by
+employing this, users can avoid using key prefixes to create an object in Cache.
+
+== Public API
+
+The work is for the new API only. The Hash data structure has the format of:
+
+[source,java]
+.Hash.java
+----
+public interface Hash<K, V> { // <1>
+
+   String getName();
+
+   V get(K key);
+
+   V put(K key, V value);
+
+   V remove(K key);
+
+   boolean containsKey(K key);
+
+   Set<K> keySet();
+
+   Collection<V> values();
+
+   Set<Map.Entry<K, V>> entrySet();
+
+   void clear();
+
+   int size();
+}
+----
+
+<1> The Hash data structure is a generic interface over `K` and `V`. Should we extend `Map<K, V>`?
+This way we could provide a wide range of operations over the Hash.
```

Comment:
I would avoid extending `Map` as that has a million methods that have made it a pain in the past for our `Cache`.

Reply:
yes, why not using map directly if we do that.

Reply:
Some further information:
* `Map` is a sync-only API, so this does not help with the async and mutiny variants
* Some of the `Map` methods have effects, such as returning the previous value on `put` and `remove` which are costly in remote scenarios.